### PR TITLE
refactor: Add NMS GitHub Actions Workflow

### PR DESCRIPTION
**Pull request by jwagantall**
Add initial cloud GHA workflow.
These workflow is migrated from CircleCI.

Signed-off-by: Jessica Wagantall <jwagantall@linuxfoundation.org>

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: .github/workflows/nms-workflow.yml [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+          str="$(jq '.head_commit.message' $GITHUB_EVENT_PATH)"    # get the head_commit message
+          echo ::set-output name=title::${str%%\\n*} | tr -d '"'
+      # Notify ci channel when failing
+      # Plugin info: https://github.com/marketplace/actions/slack-notify
+      # yamllint enable
+      - name: Notify failure to slack
+        if: failure() && github.ref == 'refs/heads/master'
+        uses: rtCamp/action-slack-notify@v2.2.0
+        env:
+          SLACK_WEBHOOK: ${{ secrets.SLACK_WEBHOOK_CI }}
```
Comment (quentinDERORY) [inline .github/workflows/nms-workflow.yml]: Use SLACK_WEBHOOK here.
When you use SLACK_WEBHOOK_CI it notifies ci-info so modify the comments too.

File: .github/workflows/nms-workflow.yml [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+          SLACK_FOOTER: ' '
+  nms-build:
+    needs:
+      - path_filter
+      - nms-flow-test
+      - eslint
+      - nms-yarn-test
+      - nms-e2e-test
+    if: ${{ needs.path_filter.outputs.should_skip == 'false' }}
+    name: xwfm deploy job
```
Comment (quentinDERORY) [inline .github/workflows/nms-workflow.yml]: wrong name here

File: .github/workflows/nms-workflow.yml [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+          SLACK_TITLE: "NMS flow-test failed"
+          SLACK_MESSAGE: "${{ steps.commit.outputs.title}}"
+          SLACK_USERNAME: "NMS workflow"
+          SLACK_ICON_EMOJI: ":boom:"
+          SLACK_COLOR: "#FF0000"
+          SLACK_FOOTER: ' '
+  eslint:
+    needs: path_filter
+    if: ${{ needs.path_filter.outputs.should_skip == 'false' }}
+    name: eslint job
```
Comment (quentinDERORY) [inline .github/workflows/nms-workflow.yml]: Can we prefix all the job names with nms.

File: .github/workflows/nms-workflow.yml [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+      MAGMA_ROOT: "${{ github.workspace }}"
+      NMS_ROOT: "${{ github.workspace }}/nms/packages/magmalte"
+      DOCKER_REGISTRY: "${{ secrets.DOCKER_REGISTRY }}"
+      DOCKER_USERNAME: "${{ secrets.DOCKER_USERNAME }}"
+      DOCKER_PASSWORD: "${{ secrets.DOCKER_PASSWORD }}"
+    steps:
+      - uses: actions/checkout@v2
+      - name: Run docker compose
+        id: nms-docker-compose
+        continue-on-error: true
```
Comment (quentinDERORY) [inline .github/workflows/nms-workflow.yml]: We only use continue-on-error if you want to implement a retry step right after

File: .github/workflows/nms-workflow.yml (hunk 1 above)
Comment (jwagantall) [reply .github/workflows/nms-workflow.yml]: fixed!

File: .github/workflows/nms-workflow.yml (hunk 2 above)
Comment (jwagantall) [reply .github/workflows/nms-workflow.yml]: you discovered my secret! :) I used an existing template to work on top of it. :)
Fixed.

File: .github/workflows/nms-workflow.yml (hunk 3 above)
Comment (jwagantall) [reply .github/workflows/nms-workflow.yml]: sure!

File: .github/workflows/nms-workflow.yml (hunk 4 above)
Comment (jwagantall) [reply .github/workflows/nms-workflow.yml]: fixed

File: .github/workflows/nms-workflow.yml [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    steps:
+      - uses: actions/checkout@v2
+      - name: Run docker compose
+        id: nms-docker-compose
+        # yamllint disable rule:line-length
+        run: |
+          cd ${MAGMA_ROOT}/nms/packages/magmalte
+          docker-compose build magmalte
+      - name: Tag and push to Docker Registry
+        if: github.ref == 'refs/heads/master' && github.event_name != 'pull_request'
```
Comment (quentinDERORY) [inline .github/workflows/nms-workflow.yml]: can you replace` github.event_name != 'pull_request'` with  `github.event_name =='push'`

Comment (quentinDERORY) [reply .github/workflows/nms-workflow.yml]: and everywhere else we use this

File: .github/workflows/nms-workflow.yml [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        # yamllint disable rule:line-length
+        run: |
+            ./ci-scripts/tag-push-docker.sh --images 'magmalte' --tag "${GITHUB_SHA:0:8}" --tag-latest false --project magmalte
+      - name: Tag and push to Jfrog Registry
+        if: github.ref == 'refs/heads/master'
```
Comment (quentinDERORY) [inline .github/workflows/nms-workflow.yml]: can you add the github event if condition here too

File: .github/workflows/nms-workflow.yml [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        env:
+          DOCKER_REGISTRY: "${{ secrets.JFROG_DOCKER_ORC8R_REGISTRY }}"
+          DOCKER_USERNAME: "${{ secrets.JFROG_USERNAME }}"
+          DOCKER_PASSWORD: "${{ secrets.JFROG_PASSWORD }}"
+        run: |
+            ./ci-scripts/tag-push-docker.sh --images 'magmalte' --tag "${GITHUB_SHA:0:8}" --tag-latest true --project magmalte
+      - name: Extract commit title
+        id: commit
+        if: github.ref == 'refs/heads/master'
```
Comment (quentinDERORY) [inline .github/workflows/nms-workflow.yml]: On all extract commit title step and slack notification:
Could you replace `github.ref == 'refs/heads/master'` with `github.event_name =='push'` .

File: .github/workflows/nms-workflow.yml (hunk 5 above)
Comment (jwagantall) [reply .github/workflows/nms-workflow.yml]: sure!.. i think this was the only place it was used on this workflows.

File: .github/workflows/nms-workflow.yml (hunk 6 above)
Comment (jwagantall) [reply .github/workflows/nms-workflow.yml]: absolutely!

File: .github/workflows/nms-workflow.yml (hunk 7 above)
Comment (jwagantall) [reply .github/workflows/nms-workflow.yml]: sure!

**Comment by jwagantall** [comment]
@quentinDERORY  I am stuck on this test
https://github.com/magma/magma/runs/3669897878?check_suite_focus=true

Is this a real failure?

**Comment by jwagantall** [comment]
> @quentinDERORY I am stuck on this test
> https://github.com/magma/magma/runs/3669897878?check_suite_focus=true
> 
> Is this a real failure?

Nevermind!.. i fixed it :)

**Comment by jwagantall** [comment]
@quentinDERORY .. I feel i have seen this before
https://github.com/magma/magma/runs/3670150182?check_suite_focus=true

Do you remember?

**Comment by quentinDERORY** [comment]
> @quentinDERORY .. I feel i have seen this before
> https://github.com/magma/magma/runs/3670150182?check_suite_focus=true
> 
> Do you remember?

yes this has to do with the -t option in docker. Here is the change we did last time https://github.com/magma/magma/pull/7974/files

But looks like in your case, you are missing the secret files... just a warning, this is happening in circle ci too https://app.circleci.com/pipelines/github/magma/magma/33295/workflows/f64e9149-d658-479b-af72-73992f872228/jobs/388341

**Comment by jwagantall** [comment]
> > @quentinDERORY .. I feel i have seen this before
> > https://github.com/magma/magma/runs/3670150182?check_suite_focus=true
> > Do you remember?
> 
> yes this has to do with the -t option in docker. Here is the change we did last time https://github.com/magma/magma/pull/7974/files
> 
> But looks like in your case, you are missing the secret files... just a warning, this is happening in circle ci too https://app.circleci.com/pipelines/github/magma/magma/33295/workflows/f64e9149-d658-479b-af72-73992f872228/jobs/388341

right!.. i remember now.. !
I noticed we are not running docker with -t in this case.
Can anyone help us adding the missing secret files?

Thanks so much!

**Comment by quentinDERORY** [comment]
> > > @quentinDERORY .. I feel i have seen this before
> > > https://github.com/magma/magma/runs/3670150182?check_suite_focus=true
> > > Do you remember?
> > 
> > 
> > yes this has to do with the -t option in docker. Here is the change we did last time https://github.com/magma/magma/pull/7974/files
> > But looks like in your case, you are missing the secret files... just a warning, this is happening in circle ci too https://app.circleci.com/pipelines/github/magma/magma/33295/workflows/f64e9149-d658-479b-af72-73992f872228/jobs/388341
> 
> right!.. i remember now.. !
> I noticed we are not running docker with -t in this case.
> Can anyone help us adding the missing secret files?
> 
> Thanks so much!

it's working without them in circle ci so no need to add them.
just try to add the docker options in the script where it uses docker exec

**Comment by jwagantall** [comment]
@quentinDERORY  is this timeout something caused by my executors?
https://github.com/magma/magma/runs/3679881137?check_suite_focus=true

**Comment by quentinDERORY** [comment]
> @quentinDERORY is this timeout something caused by my executors?
> https://github.com/magma/magma/runs/3679881137?check_suite_focus=true

Might be the case, could you try increasing the timeouts in the testcases? In the meantime you can remove the requirement to launch the build job just to test it out.

**Comment by jwagantall** [comment]
> > @quentinDERORY is this timeout something caused by my executors?
> > https://github.com/magma/magma/runs/3679881137?check_suite_focus=true
> 
> Might be the case, could you try increasing the timeouts in the testcases? In the meantime you can remove the requirement to launch the build job just to test it out.

Thnaks!.. i can confirm that the build job works 
https://github.com/magma/magma/runs/3692732582?check_suite_focus=true

**Comment by quentinDERORY** [comment]
> @quentinDERORY is this timeout something caused by my executors?
> https://github.com/magma/magma/runs/3679881137?check_suite_focus=true

it's not a resource constraint, as the same job is now failing in circle ci too. I think adding the -i created a regression

**Review by andreilee — CHANGES_REQUESTED** [review]
Looks good overall, though if you can keep the circle nms-e2e-tests from failing that would be ideal.

File: .github/workflows/nms-workflow.yml [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+        with:
+          cmd: install  # will run `yarn install` command
+      - name: run e2e_test_setup.sh
+        run: |
+            source $NVM_DIR/nvm.sh
+            nvm install stable
+            cd ${MAGMA_ROOT}/nms/packages/magmalte
+            ./e2e_test_setup.sh
+      - name: Publish Unit Test Results
+        uses: EnricoMi/publish-unit-test-result-action@v1
```
Comment (quentinDERORY) [inline .github/workflows/nms-workflow.yml]: we should use the unit-test-workflow to publish the test results as the pull request event will not have write permission.
for example see here https://github.com/magma/magma/blob/9c186ea252172ed32036f4e08e52c1772d0e8926/.github/workflows/cloud-workflow.yml#L161 and here https://github.com/magma/magma/blob/master/.github/workflows/unit-test-workflow.yml

Comment (jwagantall) [reply .github/workflows/nms-workflow.yml]: thank you!.. fixed

Comment (quentinDERORY) [reply .github/workflows/nms-workflow.yml]: you did not added the nms workflow to the trigger list in there https://github.com/magma/magma/blob/master/.github/workflows/unit-test-workflow.yml

Comment (jwagantall) [reply .github/workflows/nms-workflow.yml]: thanks for catching this!.. i missed it

**Review by quentinDERORY — APPROVED** [review]

**Pull request merged**